[AMDGPU] Do not fold an immediate into instructions with frame indexes

changpeng opened the pull request:
Do not fold an immediate into an instruction that already has a frame index operand.
A frame index could possibly turn out to be another immediate.

Fixes: SWDEV-536263

changpeng committed:
[AMDGPU] Do not fold an immediate into instructions with frame indexes

  Do not fold an immediate into an instruction that already has a frame
index operand. A frame index could possibly turn out to be another
immediate.

arsenm reviewed (commented):
Missing a new test that would break. This looks too conservative; it will always be OK if the immediate is an inline immediate

changpeng committed:
[AMDGPU] Add the new test that fails otherwise

changpeng commented:
> Missing a new test that would break. This looks too conservative; it will always be OK if the immediate is an inline immediate

It is filtered out under the condition  (MO) !IsInlineConst.

changpeng commented:
> Missing a new test that would break. This looks too conservative; it will always be OK if the immediate is an inline immediate

new test added

cdevadas commented on llvm/test/CodeGen/AMDGPU/GlobalISel/flat-scratch.ll:
```diff
@@ -1933,7 +1934,8 @@ define amdgpu_kernel void @store_load_large_imm_offset_kernel() {
 ; GFX10-NEXT:    s_setreg_b32 hwreg(HW_REG_FLAT_SCR_HI), s9
 ; GFX10-NEXT:    v_mov_b32_e32 v0, 13
 ; GFX10-NEXT:    v_mov_b32_e32 v1, 15
-; GFX10-NEXT:    s_movk_i32 s0, 0x3e84
```
The codegen looks better in the original case. It is true for most of the lit-test changes introduced in this patch.
Was this move originally incorrect? Or this code bloat is a side-effect of this fix?

changpeng replied:
> The codegen looks better in the original case. It is true for most of the lit-test changes introduced in this patch. Was this move originally incorrect? Or this code bloat is a side-effect of this fix?

If an instruction has both FI and imm operands, it could possibly result in two immediate. So this is to fix the correctness issue conservatively.

cdevadas replied:
Ok.

changpeng commented:
> Missing a new test that would break. This looks too conservative; it will always be OK if the immediate is an inline immediate

Hi, @arsenm :  New test has been added. And we already filtered out the case of inline immediate, i.e. we can fold frame index into an instruction with an inline immediate operand.
Do you have any further suggestion?

Also do you think it is cheaper just to revert the offending patch: https://github.com/llvm/llvm-project/pull/140587

arsenm commented on llvm/test/CodeGen/AMDGPU/no-folding-imm-to-inst-with-fi.ll:
```diff
@@ -0,0 +1,114 @@
+bb:
+  %alloca = alloca <4 x i64>, align 32, addrspace(5)
+  %alloca1 = alloca <16 x i64>, align 128, addrspace(5)
+  %addrspacecast = addrspacecast ptr addrspace(5) %alloca to ptr
```
Are these casts really necessary?

changpeng commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -6120,10 +6120,11 @@ bool SIInstrInfo::isOperandLegal(const MachineInstr &MI, unsigned OpIdx,
+      // already has a frame index. The frame index handling code doesn't handle
+      // fixing up operand constraints if there are multiple indexes, and a
+      // frame index could possibly turn out to be another immediate.
+      if (Op.isFI() && (MO->isFI() || MO->isImm()))
```
> This probably should be !isReg, the same issue probably exists for any exotic operand types
we are under the following conditions. So it means completely giving up when the instruction has a FI operand.
 if (!IsInlineConst && !MO->isReg() && isSALU(MI)) {

changpeng committed:
[AMDGPU] Update the patch based on the comments

  1) extend the restriction to all !isReg: the same issue probably
     exists for any exotic operand types.
  2) Update the newly added LIT test:
     -- remove unnecessary addrespacecasts
     -- store value instead of posion

changpeng committed:
[AMDGPU] fix a typo in comment

changpeng committed:
Update llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>
